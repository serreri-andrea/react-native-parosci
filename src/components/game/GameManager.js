/**
 * Created by Andrea on 18/03/2018.
 */

import React, { Component }     from 'react';
import {
    StyleSheet,
    Text,
    View,
    Button,
    BackHandler,
    Alert,
} from 'react-native';
import ScoreBoard               from "./ScoreBoard";
import CardBoard                from "./CardBoard";
import IAManager                from "./IAManager";
import GameRuler                from "../../config/game";
import Rules                    from "../../lib/Rules";
import Answers                  from "../../lib/Answers";

export default class GameManager extends Component{

    constructor(props) {
        super(props);
        this.turn = 0;
        this.previousCard = null;
        this.state = {
            p1Score:0,
            p2Score:0,
        }
    }

    componentDidMount()  {
        /* BackHandler.addEventListener('hardwareBackPress', function() {
             this.props.callback(false)
             return true;
         });*/
    }

    componentWillUnmount(){
    }

    handleBack(){
        Alert.alert(
            "Warning",
            "Do you want to exit the game ?",
            [
                {text: "yes", onPress: () => this.props.updateGame()},
                {text: "no", onPress:null}
            ],
            {cancelable: false}
        )
    }

    updateScore(cardA, cardB){
        let winner = Rules.getVictory(this.props.mode, cardA,cardB);
        let p1Score= this.state.p1Score;
        let p2Score = this.state.p2Score;
        if (winner === 1){
            this.setState({p1Score: p1Score + 1})
        }else if (winner === 2){
            this.setState({p2Score: p2Score + 1})
        }else{
            this.setState({p1Score: p1Score, p2Score: p2Score})
        }
    }

    getResponse(card){
        this.turn = this.turn + 1;
        let difficulty = this.props.difficulty;
        let mode = this.props.mode;
        this.previousCard = this.state.playerOneCard;
        if (difficulty === "medium"){
            let IACard = Answers.answersIAMedium(mode, this.previousCard, card);
            this.updateScore(card, IACard);
            this.setState({playerTwoCard:IACard})
        }else if (difficulty === "hard"){
            let IACard = Answers.answersIACheat(mode, card);
            this.updateScore(card, IACard);
            this.setState({playerTwoCard:IACard})
        }else{
            let IACard = Answers.answersIAEasy(mode);
            this.updateScore(card, IACard);
            this.setState({playerTwoCard:IACard})
        }
        this.setState({playerOneCard:card, card:null})
    }

    renderGame(){
        if (this.props.type === "pve") {
            if (this.state.p1Score >= 3 || this.state.p2Score >= 3) {
                return (
                    <View>
                        <Text> game if finished with, p1 score : {this.state.p1Score} and p2
                            score: {this.state.p2Score}</Text>
                    </View>
                )
            } else {
                return (
                    <CardBoard cards={Rules.getCards(this.props.mode)} callback={this.getResponse.bind(this)}/>
                )
            }
        }else{
            if (this.state.p1Score > 3 || this.state.p2Score > 3) {
                return(
                    <View>
                        <Text> game if finished with, IA 1 : {this.state.p1Score} and IA 2 score: {this.state.p2Score}</Text>
                    </View>
                )
            }else{
                return(
                    <IAManager cards={Rules.getCards(this.props.mode)} mode={this.props.mode} callback={this.getResponse.bind(this)} score={{p1Score:this.state.p1Score, p2Score:this.state.p2Score}}/>

                )
            }
        }
    }

    render(){
        return(
            <View style={styles.container}>
                <Text> TOUR : {this.turn} </Text>
                <ScoreBoard playerOneCard={this.state.playerOneCard} playerTwoCard={this.state.playerTwoCard}
                            mode={this.props.mode} score={{p1Score:this.state.p1Score, p2Score:this.state.p2Score}}/>
                {this.renderGame()}
                <Button
                    title="back"
                    color="pink"
                    onPress={this.handleBack.bind(this)}/>
            </View>
        )
    }
}

const styles = StyleSheet.create({
    container:{
        flex:1,
        backgroundColor:"white"
    },
});
