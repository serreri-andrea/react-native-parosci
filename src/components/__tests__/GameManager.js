import 'react-native';
import React            from 'react';
import GameManager      from '../game/GameManager';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

it('renders correctly: without props', () => {
    const tree = renderer.create(
        <GameManager/>
    );
    expect(tree).toMatchSnapshot();
});

it('renders correctly: without props', () => {
    const tree = renderer.create(
        <GameManager type={"hytgdefesefre"} mode={"fregfregregregre"} difficulty={"dedededede"} updateGame={42}/>
    );
    expect(tree).toMatchSnapshot();
});



