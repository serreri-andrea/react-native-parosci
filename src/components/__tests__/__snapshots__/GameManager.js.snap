// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly: without props 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ececec",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Image
      source={1}
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "height": undefined,
            "width": undefined,
          },
          undefined,
        ]
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#00BFA5",
          "fontSize": 20,
          "textAlign": "center",
        }
      }
    >
      Turn
      : 
      0
       
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "white",
            "fontSize": 18,
          }
        }
      >
        Score
        :
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#E53935",
              "fontSize": 18,
            }
          }
        >
           
          0
           - 
          0
        </Text>
      </Text>
      <View />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    >
      <RCTScrollView
        ItemSeparatorComponent={[Function]}
        columnWrapperStyle={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "marginBottom": 10,
            "marginTop": 10,
          }
        }
        data={
          Array [
            Object {
              "images": Object {
                "default": 1,
                "selected": 1,
              },
              "key": "rock",
              "loose": Array [
                "paper",
              ],
              "reference": "rock",
              "wins": Array [
                "scissors",
              ],
            },
            Object {
              "images": Object {
                "default": 1,
                "selected": 1,
              },
              "key": "paper",
              "loose": Array [
                "scissors",
              ],
              "reference": "paper",
              "wins": Array [
                "rock",
              ],
            },
            Object {
              "images": Object {
                "default": 1,
                "selected": 1,
              },
              "key": "scissors",
              "loose": Array [
                "rock",
              ],
              "reference": "scissors",
              "wins": Array [
                "paper",
              ],
            },
          ]
        }
        disableVirtualization={false}
        extraData={Object {}}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        numColumns={3}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                  },
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                    "marginBottom": 10,
                    "marginTop": 10,
                  },
                ]
              }
            >
              <View>
                <Image
                  source={1}
                  style={
                    Object {
                      "height": 166.75,
                      "margin": 10,
                      "width": 150,
                    }
                  }
                />
              </View>
              <View>
                <Image
                  source={1}
                  style={
                    Object {
                      "height": 166.75,
                      "margin": 10,
                      "width": 150,
                    }
                  }
                />
              </View>
              <View>
                <Image
                  source={1}
                  style={
                    Object {
                      "height": 166.75,
                      "margin": 10,
                      "width": 150,
                    }
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
    <View
      accessibilityComponentType="button"
      accessibilityLabel={undefined}
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={undefined}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
              Object {
                "color": "#E53935",
              },
            ]
          }
        >
          Return to menu
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`renders correctly: without props 2`] = `
<View
  style={
    Object {
      "backgroundColor": "#ececec",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <Image
      source={1}
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "height": undefined,
            "width": undefined,
          },
          undefined,
        ]
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#00BFA5",
          "fontSize": 20,
          "textAlign": "center",
        }
      }
    >
      Turn
      : 
      0
       
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "white",
            "fontSize": 18,
          }
        }
      >
        Score
        :
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#E53935",
              "fontSize": 18,
            }
          }
        >
           
          0
           - 
          0
        </Text>
      </Text>
      <View />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginBottom": 10,
          "marginTop": 10,
        }
      }
    >
      <RCTScrollView
        ItemSeparatorComponent={[Function]}
        columnWrapperStyle={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "marginBottom": 10,
            "marginTop": 10,
          }
        }
        data={
          Array [
            Object {
              "images": Object {
                "default": 1,
                "selected": 1,
              },
              "key": "rock",
              "loose": Array [
                "paper",
              ],
              "reference": "rock",
              "wins": Array [
                "scissors",
              ],
            },
            Object {
              "images": Object {
                "default": 1,
                "selected": 1,
              },
              "key": "paper",
              "loose": Array [
                "scissors",
              ],
              "reference": "paper",
              "wins": Array [
                "rock",
              ],
            },
            Object {
              "images": Object {
                "default": 1,
                "selected": 1,
              },
              "key": "scissors",
              "loose": Array [
                "rock",
              ],
              "reference": "scissors",
              "wins": Array [
                "paper",
              ],
            },
          ]
        }
        disableVirtualization={false}
        extraData={Object {}}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        numColumns={3}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                  },
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                    "marginBottom": 10,
                    "marginTop": 10,
                  },
                ]
              }
            >
              <View>
                <Image
                  source={1}
                  style={
                    Object {
                      "height": 166.75,
                      "margin": 10,
                      "width": 150,
                    }
                  }
                />
              </View>
              <View>
                <Image
                  source={1}
                  style={
                    Object {
                      "height": 166.75,
                      "margin": 10,
                      "width": 150,
                    }
                  }
                />
              </View>
              <View>
                <Image
                  source={1}
                  style={
                    Object {
                      "height": 166.75,
                      "margin": 10,
                      "width": 150,
                    }
                  }
                />
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
    <View
      accessibilityComponentType="button"
      accessibilityLabel={undefined}
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={undefined}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
              Object {
                "color": "#E53935",
              },
            ]
          }
        >
          Return to menu
        </Text>
      </View>
    </View>
  </View>
</View>
`;
